### ISSUE TITLE
Allow functions to be mutated

### ISSUE THREAD

**Issue by BackEndTea** [infection/infection#1482]
**Is your feature request related to a problem? Please describe.**

Right now ( due to autoloading), infection can not mutate functions, but it can mutate methods.
e.g
```php
function add(int $a, int $b): int 
{
  // no mutation here
  return $a + $b;
}

class Calculator
{
  function add(int $a, int $b): int 
  {
    // + is mutated to -
    return $a + $b;
  }
}
```

Currently the Interceptor is registered in the bootstrap file, so it will work as follows:

* `vendor/bin/phpunit`
*  autoloader gets registered
    *  all functions are autoloaded
* phpunit executes bootstrap file
    * interceptor is registered
* Interceptor intercepts autoloader

**Describe the solution you'd like**

If we move the registering of the interceptor before the autoloader is registered, we can mutate functions, as we intercept the autoloading before they are autoloaded.

We could have our own php script, instead of running the original one, like so:

```php
<?php

require_once "path_to_interceptor";

IncludeInterceptor::intercept('{$originalFilePath}', '{$mutantFilePath}');
IncludeInterceptor::enable();

require_once "original_executable"
```

**Describe alternatives you've considered**
N/A

**Additional context**

**Comment by vudaltsov** [comment]
Today I've suddenly realized that infection does not work with functions... Looking forward to this feature!

**Comment by maks-rafalko** [comment]
There was a [WIP PR](https://github.com/infection/infection/issues/1578) back in 2021, but nobody has tried to finish it since then.

Help is welcome :)

@sivlev @vudaltsov what are your use-cases, just for understanding. I guess `thecodingmachine/safe` is an example of whre this could be useful.

**Comment by vudaltsov** [comment]
> [@vudaltsov](https://github.com/vudaltsov) what are your use-cases? Just for understanding. I guess `thecodingmachine/safe` is an example of a package this could be useful.

Hi, Maks! I use functions instead of static class helpers, because they look cleaner in code. I know that functions cannot be autoloaded yet, however with non-dying runtimes it does not affect performance at all. So in some of my open source libs I might have 25% of code in functions.

For example, for `0.5` version of [typhoon/types](https://github.com/typhoon-php/type) I refactored `types::XXX()` factories to functions and constants: https://github.com/typhoon-php/type/blob/0.5.x/src/types.php .

**Comment by sanmai** [comment]
I can see one problem with functions: for Infection to mutate functions it need to be able to unload them and reload them from a different location. But this directive mandates that functions are included at all times, so there's no opportunity for Infection to re-include mutated versions of the functions. I hope you can see the predicament here. 

https://github.com/typhoon-php/type/blob/fa998b6cce36150cc41ad62fdc52163860a6a0ae/composer.json#L34-L36

**Comment by vudaltsov** [comment]
> I can see one problem with functions: for Infection to mutate functions it need to be able to unload them and reload them from a different location. But this directive mandates that functions are included at all times, so there's no opportunity for Infection to re-include mutated versions of the functions. I hope you can see the predicament here.
> 
> https://github.com/typhoon-php/type/blob/fa998b6cce36150cc41ad62fdc52163860a6a0ae/composer.json#L34-L36

What if Infection only supports mutation of functions wrapped in `if (!function_exists('xxx')) { function xxx() {} }` checks?

**Comment by sanmai** [comment]
Well, it is hard to expect that every other project will follow this pattern. And if for some reason they don't, it is going to be a fatal error with no mutations whatsoever. If I may suggest that you call static functions defined elsewhere from your global functions, then no changes to Infection will be necessary.

**Comment by vudaltsov** [comment]
Here's the hack that will actually disable a file registered via `autoload.files`:

```php
<?php

function disableAutoloadFile(string $vendorDir, string $file): void
{
    $file = realpath($file);
    assert($file !== false, sprintf('Failed to realpath file `%s`', $file));

    /** @var array<non-empty-string, non-empty-string> */
    $hashToFileMap = require $vendorDir . '/composer/autoload_files.php';
    $hash = array_search($file, $hashToFileMap);

    if (!is_string($hash)) {
        throw new LogicException(sprintf(
            'File `%s` is not registered in the `autoload.files` section of composer.json',
            $file,
        ));
    }

    if (isset($GLOBALS['__composer_autoload_files'][$hash])) {
        throw new LogicException(sprintf('File `%s` is already loaded', $file));
    }

    $GLOBALS['__composer_autoload_files'][$hash] = true;
}

disableAutoloadFile('src/functions.php');

require_once 'tmp/functions_mutated.php';
```

@maks-rafalko, if you like this "hack", I will be happy to work on the PR.

**Comment by sanmai** [comment]
Infection does something more complicated than a simple require. I encourage you to investigate it further with hopes that you will see firsthand that switching to static functions is far easier. But what do I know: we could be looking at a new amazing feature. For you, the sky is the limit. Go for it, ask for permission later.

**Comment by maks-rafalko** [comment]
> [@maks-rafalko](https://github.com/maks-rafalko), if you like this "hack", I will be happy to work on the PR.

absolutely! Comparing to other things we do here in Infection (e.g. [here](https://github.com/infection/include-interceptor)) it's not even a hack ;) Looking forward to your PR, happy to help with anything I can.

**Comment by vudaltsov** [comment]
Today I spent 2 hours exploring Infection's source code. Given that I use Infection quite often, I found it readable and clear. Great job, maintainers!

I realized that `IncludeInterceptor` already works perfectly for functions — no extra hacks needed. Next time, I'll tweak `ReflectionVisitor` to support functions and see what else needs to be done.
